keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,1169 @@
+on the server might still be running after the timeout happens.
+
+If the request handler on the server continues to write to the underlying
+`ResponseWriter` object, the `Write` method will eventually return an
+`i/o timeout` error (as soon as the internal buffer of the `ResponseWriter` 
+object is full).
+
+The `FlushError` method, on the other hand, is expected to return
+an `i/o timeout` error immediately.
+
+[1] https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/h2_bundle.go#L6570-L6594
+
+[2] https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/h2_bundle.go#L5691-L5699
+
+
+##### Read Deadline
+
+`SetReadDeadline` sets up a `time.AfterFunc` [1] with the given deadline, once 
+the read deadline exceeds, it closes the request `Body` with
+an `i/o timeout` error [2].
+
+> Note that in the case of read deadline exceeding, the stream is not closed with a `stream reset` error
+
+
+[1] https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/h2_bundle.go#L6544-L6568
+
+[2] https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/h2_bundle.go#L5681-L5689
+
+
+##### Integrate FlushError
+
+The `Flush` method of the underlying `ResponseWriter` object does not return an
+`error`. With the introduction of per-request read/write deadline the golang team
+has added a new version of `Flush` that returns an `error`, for both http/1x and http/2.0
+
+```
+FlushError() error
+```
+
+[1] https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/server.go#L1719C20-L1729
+
+[2] https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/h2_bundle.go#L6600-L6623
+
+
+We need to replace the use of `Flush` with `FlushError`:
+```
+// Flush
+flusher, ok := w.(http.Flusher)
+flusher.Flush()
+
+// Replace with FlushError
+flusher, ok := w.(interface{ FlushError() error })
+if err := flusher.FlushError(); err != nil {
+	return err
+}
+```
+
+> With the use of `FlushError`, the request handler can terminate as soon as it 
+> receives an error from `FlushError` after the timeout happens.
+
+
+The following code sites use `Flush` today:
+- https://github.com/kubernetes/kubernetes/blob/ba57653b676fa639f86345f2bca20bc9e1bdc8a9/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go#L194
+- https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/staging/src/k8s.io/apiserver/pkg/util/flushwriter/writer.go#L50
+
+
+### Termination of the Request Handler
+
+Upon timeout, a request handler can self-terminate by returning the error it receives while:
+- reading from the `Body` of the Request
+- writing via the `Write` method of the `ResponseWriter` object, or
+- flushing using the `FlushError` method
+
+
+We can wrap the `ResponseWriter`, and the `Body` of the request to panic with an
```

Comment:
for alpha the entirety of the proposed feature will be behind the feature flag `PerHandlerReadWriteTimeout`. 
I am trying to understand what it means for an admin: after enabling the proposed feature, the admin will also want to be able to configure whether the inner handler panics or errors?

I don't think the proposed panic behavior can be enabled independently, unless we say that we want to change the current timeout handler to throw a panic instead of error and do this behind a feature gate, thoughts?

Reply:
I added the additional feature gate for the panic behavior